File: examples/typescript/src/common/rpc.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import * as dotenv from "dotenv";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export function findRepoRoot(): string {
  let current = __dirname;
  while (current !== path.dirname(current)) {
    if (
      fs.existsSync(path.join(current, ".git")) ||
      fs.existsSync(path.join(current, ".env.example"))
    ) {
      return current;
    }
    current = path.dirname(current);
  }
  return process.cwd();
}

export interface SandboxConfig {
  rpcUrl: string;
  rpcUser: string;
  rpcPassword: string;
  rpcWallet: string;
  zmqRawBlock: string;
  zmqRawTx: string;
}

export function getConfig(): SandboxConfig {
  const repoRoot = findRepoRoot();
  const envPath = path.join(repoRoot, ".env");
  const envExamplePath = path.join(repoRoot, ".env.example");

  if (fs.existsSync(envPath)) {
    dotenv.config({ path: envPath });
  } else if (fs.existsSync(envExamplePath)) {
    dotenv.config({ path: envExamplePath });
  }

  return {
    rpcUrl: process.env.BITCOIN_RPC_URL || "http://127.0.0.1:18443",
    rpcUser: process.env.BITCOIN_RPC_USER || "bitcoinrpc",
    rpcPassword: process.env.BITCOIN_RPC_PASSWORD || "bitcoinrpcpassword",
    rpcWallet: process.env.BITCOIN_RPC_WALLET || "lab",
    zmqRawBlock: process.env.BITCOIN_ZMQ_RAWBLOCK || "tcp://127.0.0.1:28332",
    zmqRawTx: process.env.BITCOIN_ZMQ_RAWTX || "tcp://127.0.0.1:28333",
  };
}

export class BitcoinRPC {
  public baseUrl: string;
  public user: string;
  public password: string;
  public wallet?: string;

  constructor(options?: {
    url?: string;
    user?: string;
    password?: string;
    wallet?: string;
  }) {
    const config = getConfig();
    this.baseUrl = (options?.url || config.rpcUrl).replace(/\/+$/, "");
    this.user = options?.user || config.rpcUser;
    this.password = options?.password || config.rpcPassword;
    this.wallet = options?.wallet;
  }

  get url(): string {
    if (this.wallet) {
      return `${this.baseUrl}/wallet/${this.wallet}`;
    }
    return this.baseUrl;
  }

  async call<T = any>(method: string, params: any[] = []): Promise<T> {
    const authHeader =
      "Basic " + Buffer.from(`${this.user}:${this.password}`).toString("base64");

    let response: Response;
    try {
      response = await fetch(this.url, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: authHeader,
        },
        body: JSON.stringify({
          jsonrpc: "1.0",
          id: "bitcoin-sandbox-ts",
          method,
          params,
        }),
      });
    } catch (err: any) {
      throw new Error(
        `Could not connect to Bitcoin node at ${this.url}. Is bitcoind running? (Run: bash scripts/init-lab.sh)`
      );
    }

    const data = (await response.json()) as any;
    if (data.error) {
      throw new Error(
        `Bitcoin RPC Error (${data.error.code}): ${data.error.message}`
      );
    }

    return data.result as T;
  }

  async getRest<T = any>(endpoint: string): Promise<T> {
    const cleanEndpoint = endpoint.replace(/^\/+/, "");
    const url = `${this.baseUrl}/rest/${cleanEndpoint}`;

    let response: Response;
    try {
      response = await fetch(url);
    } catch (err: any) {
      throw new Error(`Could not connect to Bitcoin REST endpoint at ${url}`);
    }

    if (!response.ok) {
      throw new Error(`REST request failed with status: ${response.status}`);
    }

    return (await response.json()) as T;
  }

  forWallet(walletName: string): BitcoinRPC {
    return new BitcoinRPC({
      url: this.baseUrl,
      user: this.user,
      password: this.password,
      wallet: walletName,
    });
  }
}

File: examples/python/common/rpc.py
"""Bitcoin JSON-RPC and REST client for Python examples."""
import os
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional
import requests
from dotenv import load_dotenv


def find_repo_root() -> Path:
    """Locate the repository root directory by looking for .git or .env.example."""
    current = Path(__file__).resolve().parent
    while current != current.parent:
        if (current / ".git").exists() or (current / ".env.example").exists():
            return current
        current = current.parent
    return Path.cwd()


def get_config() -> Dict[str, str]:
    """Load configuration from environment or .env file."""
    repo_root = find_repo_root()
    env_path = repo_root / ".env"
    if env_path.exists():
        load_dotenv(dotenv_path=env_path)
    else:
        env_example = repo_root / ".env.example"
        if env_example.exists():
            load_dotenv(dotenv_path=env_example)

    return {
        "rpc_url": os.getenv("BITCOIN_RPC_URL", "http://127.0.0.1:18443"),
        "rpc_user": os.getenv("BITCOIN_RPC_USER", "bitcoinrpc"),
        "rpc_password": os.getenv("BITCOIN_RPC_PASSWORD", "bitcoinrpcpassword"),
        "rpc_wallet": os.getenv("BITCOIN_RPC_WALLET", "lab"),
        "zmq_rawblock": os.getenv("BITCOIN_ZMQ_RAWBLOCK", "tcp://127.0.0.1:28332"),
        "zmq_rawtx": os.getenv("BITCOIN_ZMQ_RAWTX", "tcp://127.0.0.1:28333"),
    }


class BitcoinRPC:
    """Client for Bitcoin Core JSON-RPC and REST endpoints."""

    def __init__(
        self,
        url: Optional[str] = None,
        user: Optional[str] = None,
        password: Optional[str] = None,
        wallet: Optional[str] = None,
    ):
        config = get_config()
        self.base_url = (url or config["rpc_url"]).rstrip("/")
        self.user = user or config["rpc_user"]
        self.password = password or config["rpc_password"]
        self.wallet = wallet
        self.session = requests.Session()
        self.session.auth = (self.user, self.password)

    @property
    def url(self) -> str:
        """Returns the full RPC endpoint URL including wallet context if set."""
        if self.wallet:
            return f"{self.base_url}/wallet/{self.wallet}"
        return self.base_url

    def call(self, method: str, params: Optional[List[Any]] = None) -> Any:
        """Execute a JSON-RPC method call."""
        if params is None:
            params = []
        payload = {
            "jsonrpc": "1.0",
            "id": "bitcoin-sandbox",
            "method": method,
            "params": params,
        }
        headers = {"content-type": "application/json"}
        try:
            response = self.session.post(self.url, json=payload, headers=headers, timeout=30)
        except requests.exceptions.ConnectionError as exc:
            raise ConnectionError(
                f"Could not connect to Bitcoin node at {self.url}. Is bitcoind running? (Run: bash scripts/init-lab.sh)"
            ) from exc

        if response.status_code not in (200, 500):
            response.raise_for_status()

        data = response.json()
        if data.get("error") is not None:
            err = data["error"]
            raise RuntimeError(f"Bitcoin RPC Error ({err.get('code')}): {err.get('message')}")

        return data.get("result")

    def get_rest(self, endpoint: str) -> Any:
        """Query an unauthenticated Bitcoin REST endpoint (e.g. 'chaininfo.json')."""
        clean_endpoint = endpoint.lstrip("/")
        url = f"{self.base_url}/rest/{clean_endpoint}"
        try:
            response = requests.get(url, timeout=10)
            response.raise_for_status()
            return response.json()
        except requests.exceptions.ConnectionError as exc:
            raise ConnectionError(
                f"Could not connect to Bitcoin REST at {url}."
            ) from exc

    def for_wallet(self, wallet_name: str) -> "BitcoinRPC":
        """Return a new BitcoinRPC client scoped to a specific wallet."""
        return BitcoinRPC(
            url=self.base_url,
            user=self.user,
            password=self.password,
            wallet=wallet_name,
        )
